test/ct-test-srv/main.go

```diff
@@ -27,83 +27,145 @@ type ctSubmissionRequest struct {
 
 type integrationSrv struct {
 	sync.Mutex
-	submissions     map[string]int64
+	submissions map[string]int64
+	// Hostnames where we refuse to provide an SCT. This is to exercise the code
+	// path where all CT servers fail.
+	rejectHosts map[string]bool
+	// A list of entries that we rejected based on rejectHosts.
+	rejected        []string
 	key             *ecdsa.PrivateKey
 	latencySchedule []float64
 	latencyItem     int
 }
 
-func (is *integrationSrv) handler(w http.ResponseWriter, r *http.Request) {
-	switch r.URL.Path {
-	case "/ct/v1/add-pre-chain":
-		fallthrough
-	case "/ct/v1/add-chain":
-		if r.Method != "POST" {
-			http.NotFound(w, r)
-			return
-		}
-		bodyBytes, err := ioutil.ReadAll(r.Body)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
-		}
+func readJSON(w http.ResponseWriter, r *http.Request, output interface{}) error {
+	if r.Method != "POST" {
+		return fmt.Errorf("incorrect method; only POST allowed")
+	}
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return err
+	}
 
-		var addChainReq ctSubmissionRequest
-		err = json.Unmarshal(bodyBytes, &addChainReq)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
-		}
-		if len(addChainReq.Chain) == 0 {
-			w.WriteHeader(400)
-			return
-		}
+	err = json.Unmarshal(bodyBytes, output)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-		precert := false
-		if r.URL.Path == "/ct/v1/add-pre-chain" {
-			precert = true
-		}
+func (is *integrationSrv) addChain(w http.ResponseWriter, r *http.Request) {
+	is.addChainOrPre(w, r, false)
+}
 
-		b, err := base64.StdEncoding.DecodeString(addChainReq.Chain[0])
-		if err != nil {
-			w.WriteHeader(400)
-			return
-		}
-		cert, err := x509.ParseCertificate(b)
-		if err != nil {
-			w.WriteHeader(400)
-			return
-		}
-		hostnames := strings.Join(cert.DNSNames, ",")
+// addRejectHost takes a JSON POST with a "host" field; any subsequent
+// submissions for that host will get a 400 error.
+func (is *integrationSrv) addRejectHost(w http.ResponseWriter, r *http.Request) {
+	var rejectHostReq struct {
+		Host string
+	}
+	err := readJSON(w, r, &rejectHostReq)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
 
-		is.Lock()
-		is.submissions[hostnames]++
-		is.Unlock()
+	is.Lock()
+	defer is.Unlock()
+	is.rejectHosts[rejectHostReq.Host] = true
+}
```

Comment:
If you don't call `w.Write`/`w.WriteHeader` does the `http` package properly cleanup the request?

Reply:
Surprisingly, the Go docs don't describe what happens if you forget to call `w.Write`, but I would consider failing to do so a bug. Will fix; thanks for noticing.